opentracing/tracer.h

```diff
@@ -1,322 +1,39 @@
-     * fails anyway.
-     *
-     * There's not specific reason for `Writer` being const. It's just to support
-     * the nice inline form described above. This might mean that you might have
-     * to use `mutable` in some cases. This `const` will be revised if there prove
-     * to be issues.
-     *
-     * See also `Tracer::join()`.
-     */
-    virtual Result inject(const Span & sp, const Writer & writer, std::string & error) const = 0;
+    virtual const SpanContext* extract(const Carrier& carrier) const = 0;
 
-    /**
-     * `join()` returns a `Span` instance with operation name `operationName` given
-     * a carrier reader `reader`.
-     *
-     * `join()` is responsible for extracting and joining to the trace of a `Span`
-     * instance embedded in a format-specific "carrier" object. Typically the
-     * joining will take place on the server side of an RPC boundary, but
-     * message queues and other IPC mechanisms are also reasonable places to
-     * use `join()`.
-     *
-     * OpenTracing defines an extensible set of `Reader`s (see `Reader`), and
-     * each supports potentially many carrier types.
-     *
-     *
-     *
-     * Example usage:
-     *
-     *      CarrierT carrier;
-     *      std::string error;
-     *
-     *      Span * span(0);
-     *      Tracer::Result result = globalTracer()->join(&span,
-     *                                                   "operation()",
-     *                                                   MapTextAdapter<CarrierT>(carrier),
-     *                                                   error);
-     *      if (result != Tracer::Success) { ...
-     *
-     * NOTE: All `Tracer` implementations MUST support `TextMapReader` `Reader`s.
-     *
-     * Return values:
-     *  - A successful `join()` will return `Tracer::Success` and will set `sp` to
-     *    the value of an appropriately initialized `Span` pointer.
-     *  - If there was simply no trace to join with in `carrier`, `join()`
-     *    returns `Tracer::ErrTraceNotFound` and `sp` isn't modified.
-     *  - If the `reader` is unsupported or unrecognized, `join()` returns
-     *    `Tracer::ErrUnsupportedFormat` and `sp` isn't modified.
-     *  - If there are more fundamental problems with the `reader` object,
-     *    `join()` may return `Tracer::ErrInvalidCarrier` or
-     *    `Tracer::ErrTraceCorrupted` and `sp` won't be modified.
-     *
-     *  See also `Tracer::inject()`.
-     */
-    virtual Result join(Span ** sp, const std::string & operationName, const Reader & reader, std::string & error) const = 0;
-};
-
-/**
- * This is the base interface for writing data to carriers. Not much
- * functionality here but provides through `dynamic_cast<>`ing the means
- * to check if an implementation is supported or not.
- */
-class Writer
-{
-public:
-    /**
-     * Destroys this Writer.
-     */
-    virtual ~Writer();
-};
+    virtual void cleanup(const SpanOptions* opts) = 0;
```

Comment:
I'd prefer that we leave deallocation out of the Tracer interface... seems like a preoptimization and a possible point of confusion